splitgpg2/__init__.py

```diff
@@ -749,39 +766,40 @@ def update_keygrip_map(self) -> None:
             'gpg', *self.homedir_opts(), '--list-secret-keys', '--with-colons'
         ])
         keys: List[KeyInfo] = []
-        key: Optional[KeyInfo] = None
+        primary_key: Optional[KeyInfo] = None
         subkey: Optional[SubKeyInfo] = None
         for line in out.split(b"\n"):
             fields = line.split(b":")
             if fields[0] in [b"sec", b"ssb", b""]:
                 if subkey is not None:
-                    assert key is not None, 'bad output from GnuPG'
-                    subkey.key = key
-                    key.subkeys.append(subkey)
+                    assert primary_key is not None, 'bad output from GnuPG'
+                    subkey.key = primary_key
+                    primary_key.subkeys.append(subkey)
                     subkey = None
-            if fields[0] in [b"sec", b""] and key is not None:
-                keys.append(key)
-
+            if fields[0] in [b"sec", b""] and primary_key is not None:
+                keys.append(primary_key)
             if fields[0] == b"sec":
-                key = KeyInfo()
+                primary_key = KeyInfo(fields[11])
             elif fields[0] == b"ssb":
-                subkey = SubKeyInfo()
+                assert primary_key is not None, 'subkey before primary key?'
+                subkey = SubKeyInfo(fields[11], primary_key)
+                primary_key.subkeys.append(subkey)
```

Comment:
Aren't you appending it twice now? It's already done above when finishing processing given subkey.